src/System.Memory/src/System/SpanHelpers.cs

```diff
@@ -0,0 +1,101 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Reflection;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    internal static class SpanHelpers 
+    {
+        /// <summary>
+        /// Computes "start + index * sizeof(T)", using the unsigned IntPtr-sized multiplication for 32 and 64 bits.
+        /// 
+        /// Assumptions:
+        ///     Start and index are non-negative, and already pre-validated to be within the valid range of their containing Span.
+        ///     
+        ///     If the byte length (Span.Length * sizeof(T)) does an unsigned overflow (i.e. the buffer wraps or is too big to fit within the address space),
+        ///     the behavior is undefined.
```

Comment:
I have asserts for start and index being negative (as those are already validated.)

The unsigned overflow is unvalidated at this point and they come from the application so it's not a condition I can assert on.